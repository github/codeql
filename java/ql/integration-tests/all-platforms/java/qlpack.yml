dependencies:
  codeql/java-all: '*'
  codeql/java-tests: '*'
  codeql/java-queries: '*'
