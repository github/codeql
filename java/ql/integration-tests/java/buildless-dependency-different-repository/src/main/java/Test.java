import testpkg.DepClass;
import testpkg2.DepClass2;

public class Test { 

  DepClass2 dc2 = DepClass.getDep2();

}
