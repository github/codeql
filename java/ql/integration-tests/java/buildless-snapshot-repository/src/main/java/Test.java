import testpkg.DepClass;

public class Test { 

  DepClass dc;

}
