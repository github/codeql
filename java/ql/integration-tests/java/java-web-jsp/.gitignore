#Eclipse
.classpath
.project
test-output
.settings

#IntelliJ
*.iml
*.ipr
*.iws
.idea/

#Gradle
.gradle

#Build directories
bin/
build/
target/

#Maven
*.versionsBackup