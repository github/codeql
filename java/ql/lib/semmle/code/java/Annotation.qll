/**
 * Provides classes and predicates for working with Java annotations.
 *
 * Annotations are used to add meta-information to language elements in a
 * uniform fashion. They can be seen as typed modifiers that can take
 * parameters.
 *
 * Each annotation type has zero or more annotation elements that contain a
 * name and possibly a value.
 */

import Element
import Expr
import Type
import Member
import JDKAnnotations

/** Any annotation used to annotate language elements with meta-information. */
class Annotation extends @annotation, Expr {
  /** Holds if this annotation applies to a declaration. */
  predicate isDeclAnnotation() { this instanceof DeclAnnotation }

  /** Holds if this annotation applies to a type. */
  predicate isTypeAnnotation() { this instanceof TypeAnnotation }

  /** Gets the element being annotated. */
  Element getAnnotatedElement() {
    exists(Element e | e = this.getParent() |
      if e.(Field).getCompilationUnit().fromSource()
      then
        exists(FieldDeclaration decl |
          decl.getField(0) = e and
          result = decl.getAField()
        )
      else result = e
    )
  }

  /** Gets the annotation type declaration for this annotation. */
  override AnnotationType getType() {
    result = Expr.super.getType().(Interface).getSourceDeclaration()
  }

  /** Gets the annotation element with the specified `name`. */
  AnnotationElement getAnnotationElement(string name) {
    result = this.getType().getAnnotationElement(name)
  }

  /**
   * DEPRECATED: Getting the value of _any_ annotation element is error-prone because
   * it could lead to selecting the value of the wrong element by accident (for example
   * when an annotation type is extended in the future). Prefer the predicate `getValue(string)`
   * and explicitly specify the element name. Use `getValue(_)` if it is really desired to
   * get the value of any element.
   *
   * Gets a value of an annotation element. This includes default values in case
   * no explicit value is specified. For elements with an array value type this
   * might have an `ArrayInit` as result. To properly handle array values, prefer
   * the predicate `getAnArrayValue`.
   */
  deprecated Expr getAValue() { filteredAnnotValue(this, _, result) }

  /**
   * Gets the value of the annotation element with the specified `name`.
   * This includes default values in case no explicit value is specified.
   * For elements with an array value type this might get an `ArrayInit` instance.
   * To properly handle array values, prefer the predicate `getAnArrayValue`.
   */
  Expr getValue(string name) { filteredAnnotValue(this, this.getAnnotationElement(name), result) }

  /**
   * Gets the value of the annotation element, if its type is not an array.
   * This guarantees that for consistency even elements of type array with a
   * single value have no result, to prevent accidental error-prone usage.
   */
  private Expr getNonArrayValue(string name) {
    result = this.getValue(name) and
    not this.getAnnotationElement(name).getType() instanceof Array
  }

  /**
   * If the value type of the annotation element with the specified `name` is an enum type,
   * gets the enum constant used as value for that element. This includes default values in
   * case no explicit value is specified.
   *
   * If the element value type is an enum type array, use `getAnEnumConstantArrayValue`.
   */
  EnumConstant getEnumConstantValue(string name) {
    result = this.getNonArrayValue(name).(FieldRead).getField()
  }

  /**
   * If the value type of the annotation element with the specified `name` is `String`,
   * gets the string value used for that element. This includes default values in case no
   * explicit value is specified.
   *
   * If the element value type is a string array, use `getAStringArrayValue`.
   */
  string getStringValue(string name) {
    // Uses CompileTimeConstantExpr instead of StringLiteral because this can for example
    // be a read from a final variable as well.
    result = this.getNonArrayValue(name).(CompileTimeConstantExpr).getStringValue()
  }

  /**
   * If the value type of the annotation element with the specified `name` is `int` or
   * a smaller integral type or `char`, gets the int value used for that element.
   * This includes default values in case no explicit value is specified.
   *
   * If the element value type is an `int` array or an array of a smaller integral
   * type or `char`, use `getAnIntArrayValue`.
   */
  int getIntValue(string name) {
    // Uses CompileTimeConstantExpr instead of IntegerLiteral because this can for example
    // be a read from a final variable as well.
    result = this.getNonArrayValue(name).(CompileTimeConstantExpr).getIntValue() and
    // Verify that type is integral; ignore floating point elements with IntegerLiteral as value
    this.getAnnotationElement(name).getType().hasName(["byte", "short", "int", "char"])
  }

  /**
   * If the value type of the annotation element with the specified `name` is `boolean`,
   * gets the boolean value used for that element. This includes default values in case
   * no explicit value is specified.
   */
  boolean getBooleanValue(string name) {
    // Uses CompileTimeConstantExpr instead of BooleanLiteral because this can for example
    // be a read from a final variable as well.
    result = this.getNonArrayValue(name).(CompileTimeConstantExpr).getBooleanValue()
  }

  /**
   * If the value type of the annotation element with the specified `name` is `java.lang.Class`,
   * gets the type referred to by that `Class`. This includes default values in case no explicit
   * value is specified.
   *
   * If the element value type is a `Class` array, use `getATypeArrayValue`.
   */
  Type getTypeValue(string name) {
    result = this.getNonArrayValue(name).(TypeLiteral).getReferencedType()
  }

  /** Gets the element being annotated. */
  Element getTarget() { result = this.getAnnotatedElement() }

  override string toString() { result = this.getType().getName() }

  /** This expression's Halstead ID (used to compute Halstead metrics). */
  override string getHalsteadID() { result = "Annotation" }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as an array
   * type. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression used as value.
   */
  Expr getAnArrayValue(string name) { result = this.getArrayValue(name, _) }

  /**
   * DEPRECATED: Predicate has been renamed to `getAnArrayValue`
   */
  deprecated Expr getAValue(string name) { result = this.getAnArrayValue(name) }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as an enum
   * type array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression used as value.
   */
  EnumConstant getAnEnumConstantArrayValue(string name) {
    result = this.getAnArrayValue(name).(FieldRead).getField()
  }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as a string
   * array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression used as value.
   */
  string getAStringArrayValue(string name) {
    result = this.getAnArrayValue(name).(CompileTimeConstantExpr).getStringValue()
  }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as an `int`
   * array or an array of a smaller integral type or `char`. This includes default values in case no
   * explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression used as value.
   */
  int getAnIntArrayValue(string name) {
    result = this.getAnArrayValue(name).(CompileTimeConstantExpr).getIntValue() and
    // Verify that type is integral; ignore floating point elements with IntegerLiteral as value
    this.getAnnotationElement(name).getType().hasName(["byte[]", "short[]", "int[]", "char[]"])
  }

  /**
   * Gets a value of the annotation element with the specified `name`, which must be declared as a `Class`
   * array. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be one of the
   * elements of that array. Otherwise, the result will be the single expression used as value.
   */
  Type getATypeArrayValue(string name) {
    result = this.getAnArrayValue(name).(TypeLiteral).getReferencedType()
  }

  /**
   * Gets the value at a given index of the annotation element with the specified `name`, which must be
   * declared as an array type. This includes default values in case no explicit value is specified.
   *
   * If the annotation element is defined with an array initializer, then the result will be the element
   * at the given index of that array, starting at 0. Otherwise, the result will be the single expression
   * defined for the value and the `index` will be 0.
   */
  Expr getArrayValue(string name, int index) {
    this.getType().getAnnotationElement(name).getType() instanceof Array and
    exists(Expr value | value = this.getValue(name) |
      if value instanceof ArrayInit
      then
        // TODO: Currently reports incorrect index values in some cases, see https://github.com/github/codeql/issues/8645
        result = value.(ArrayInit).getInit(index)
      else (
        index = 0 and result = value
      )
    )
  }

  override string getAPrimaryQlClass() { result = "Annotation" }
}

/** An `Annotation` that applies to a declaration. */
class DeclAnnotation extends @declannotation, Annotation { }

/** An `Annotation` that applies to a type. */
class TypeAnnotation extends @typeannotation, Annotation { }

/**
 * There may be duplicate entries in annotValue(...) - one entry for
 * information populated from bytecode, and one for information populated
 * from source. This removes the duplication.
 */
private predicate filteredAnnotValue(Annotation a, Method m, Expr val) {
  annotValue(a, m, val) and
  (sourceAnnotValue(a, m, val) or not sourceAnnotValue(a, m, _))
}

private predicate sourceAnnotValue(Annotation a, Method m, Expr val) {
  annotValue(a, m, val) and
  val.getFile().isSourceFile()
}

/** An abstract representation of language elements that can be annotated. */
class Annotatable extends Element {
  /**
   * Holds if this element has an annotation, including inherited annotations.
   * The retention policy of the annotation type is not considered.
   */
  predicate hasAnnotation() { exists(this.getAnAnnotation()) }

  /**
   * Holds if this element has a declared annotation, excluding inherited annotations.
   * The retention policy of the annotation type is not considered.
   */
  predicate hasDeclaredAnnotation() { exists(this.getADeclaredAnnotation()) }

  /**
   * Holds if this element has the specified annotation, including inherited
   * annotations. The retention policy of the annotation type is not considered.
   */
  predicate hasAnnotation(string package, string name) {
    exists(AnnotationType at | at = this.getAnAnnotation().getType() |
      at.nestedName() = name and at.getPackage().getName() = package
    )
  }

  /**
   * Gets an annotation that applies to this element, including inherited annotations.
   * The results only include _direct_ annotations; _indirect_ annotations, that is
   * repeated annotations in an (implicit) container annotation, are not included.
   * The retention policy of the annotation type is not considered.
   */
  cached
  Annotation getAnAnnotation() {
    // This predicate is overridden by Class to consider inherited annotations
    result = this.getADeclaredAnnotation()
  }

  /**
   * Gets an annotation that is declared on this element, excluding inherited annotations.
   * The retention policy of the annotation type is not considered.
   */
  Annotation getADeclaredAnnotation() { result.getAnnotatedElement() = this }

  /** Gets an _indirect_ (= repeated) annotation. */
  private Annotation getAnIndirectAnnotation() {
    // 'indirect' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
    exists(AnnotationType t, Annotation containerAnn |
      t = result.getType() and
      containerAnn = this.getADeclaredAnnotation() and
      containerAnn.getType() = t.getContainingAnnotationType()
    |
      result = containerAnn.getAnArrayValue("value")
    )
  }

  private Annotation getADeclaredAssociatedAnnotation(AnnotationType t) {
    // Direct or indirect annotation
    result.getType() = t and
    result = [this.getADeclaredAnnotation(), this.getAnIndirectAnnotation()]
  }

  private Annotation getAnAssociatedAnnotation(AnnotationType t) {
    // 'associated' as defined by https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/reflect/AnnotatedElement.html
    if exists(this.getADeclaredAssociatedAnnotation(t))
    then result = this.getADeclaredAssociatedAnnotation(t)
    else (
      // Only if neither a direct nor an indirect annotation is present look for an inherited one
      t.isInherited() and
      // @Inherited only works for classes; cast to Annotatable is necessary because predicate is private
      result = this.(Class).getASupertype().(Class).(Annotatable).getAnAssociatedAnnotation(t)
    )
  }

  /**
   * Gets an annotation _associated_ with this element, that is:
   * - An annotation directly present on this element, or
   * - An annotation indirectly present on this element (in the form of a repeated annotation), or
   * - If an annotation of a type is neither directly nor indirectly present
   *   the result is an associated inherited annotation (recursively)
   *
   * The retention policy of the annotation type is not considered.
   */
  Annotation getAnAssociatedAnnotation() { result = this.getAnAssociatedAnnotation(_) }

  /**
   * Holds if this or any enclosing `Annotatable` has a `@SuppressWarnings("<category>")`
   * annotation attached to it for the specified `category`.
   */
  predicate suppressesWarningsAbout(string category) {
    category = this.getAnAnnotation().(SuppressWarningsAnnotation).getASuppressedWarning()
    or
    this.(Member).getDeclaringType().suppressesWarningsAbout(category)
    or
    this.(Expr).getEnclosingCallable().suppressesWarningsAbout(category)
    or
    this.(Stmt).getEnclosingCallable().suppressesWarningsAbout(category)
    or
    this.(NestedClass).getEnclosingType().suppressesWarningsAbout(category)
    or
    this.(LocalClassOrInterface)
        .getLocalTypeDeclStmt()
        .getEnclosingCallable()
        .suppressesWarningsAbout(category)
    or
    this.(LocalVariableDecl).getCallable().suppressesWarningsAbout(category)
  }
}

/** An annotation type is a special kind of interface type declaration. */
class AnnotationType extends Interface {
  AnnotationType() { isAnnotType(this) }

  /** Gets the annotation element with the specified `name`. */
  AnnotationElement getAnnotationElement(string name) {
    methods(result, _, _, _, this, _) and result.hasName(name)
  }

  /** Gets an annotation element that is a member of this annotation type. */
  AnnotationElement getAnAnnotationElement() { methods(result, _, _, _, this, _) }

  /** Holds if this annotation type is annotated with the meta-annotation `@Inherited`. */
  predicate isInherited() {
    this.getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Inherited")
  }

  /** Holds if this annotation type is annotated with the meta-annotation `@Documented`. */
  predicate isDocumented() {
    this.getADeclaredAnnotation().getType().hasQualifiedName("java.lang.annotation", "Documented")
  }

  /**
   * Gets the retention policy of this annotation type, that is, the name of one of the
   * enum constants of `java.lang.annotation.RetentionPolicy`. If this annotation type
   * has no `@Retention` annotation, the result is `CLASS`.
   */
  string getRetentionPolicy() {
    if this.getADeclaredAnnotation() instanceof RetentionAnnotation
    then result = this.getADeclaredAnnotation().(RetentionAnnotation).getRetentionPolicy()
    else
      // If not explicitly specified retention is CLASS
      result = "CLASS"
  }

  /**
   * Holds if the element type is a possible target for this annotation type.
   * The `elementType` is the name of one of the `java.lang.annotation.ElementType`
   * enum constants.
   *
   * If this annotation type has no `@Target` annotation, it is considered to be applicable
   * in all declaration contexts. This matches the behavior of the latest Java versions
   * but differs from the behavior of older Java versions. This predicate must only be
   * called with names of `ElementType` enum constants; for other values it might hold
   * erroneously.
   */
  bindingset[elementType]
  predicate isATargetType(string elementType) {
    /*
     * Note: Cannot use a predicate with string as result because annotation type without
     * explicit @Target can be applied in all declaration contexts, requiring to hardcode
     * element types here; then the results could become outdated if this predicate is not
     * updated for future JDK versions, or it could have irritating results, e.g. RECORD_COMPONENT
     * for a database created for Java 8.
     *
     * Could in theory read java.lang.annotation.ElementType constants from database, but might
     * be brittle in case ElementType is not present in the database for whatever reason.
     */

    if this.getADeclaredAnnotation() instanceof TargetAnnotation
    then elementType = this.getADeclaredAnnotation().(TargetAnnotation).getATargetElementType()
    else
      /*
       * Behavior for missing @Target annotation changed between Java versions. In older Java
       * versions it allowed usage in most (but not all) declaration contexts. Then for Java 14
       * JDK-8231435 changed it to allow usage in all declaration and type contexts. In Java 17
       * it was changed by JDK-8261610 to only allow usage in all declaration contexts, but not
       * in type contexts anymore. However, during these changes javac did not always comply with
       * the specification, see for example JDK-8254023.
       *
       * For simplicity pretend the latest behavior defined by the JLS applied in all versions;
       * that means any declaration context is allowed, but type contexts (represented by TYPE_USE,
       * see JLS 17 section 9.6.4.1) are not allowed.
       */

      elementType != "TYPE_USE"
  }

  /** Holds if this annotation type is annotated with the meta-annotation `@Repeatable`. */
  predicate isRepeatable() { this.getADeclaredAnnotation() instanceof RepeatableAnnotation }

  /**
   * If this annotation type is annotated with the meta-annotation `@Repeatable`,
   * gets the annotation type which acts as _containing annotation type_.
   */
  AnnotationType getContainingAnnotationType() {
    result = this.getADeclaredAnnotation().(RepeatableAnnotation).getContainingType()
  }
}

/** An annotation element is a member declared in an annotation type. */
class AnnotationElement extends Member {
  AnnotationElement() { isAnnotElem(this) }

  /** Gets the type of this annotation element. */
  Type getType() { methods(this, _, _, result, _, _) }

  /** Gets the Kotlin type of this annotation element. */
  KotlinType getKotlinType() { methodsKotlinType(this, result) }
}
