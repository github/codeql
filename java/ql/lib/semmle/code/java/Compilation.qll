/**
 * Provides a class representing individual compiler invocations that occurred during the build.
 */

import semmle.code.FileSystem

/**
 * An invocation of the compiler. Note that more than one file may be
 * compiled per invocation. For example, this command compiles three
 * source files:
 *
 *   javac Foo.java Bar.java Baz.java
 *
 * Two things happen to each file during a compilation:
 *
 *   1. The file is compiled by a real compiler, such as javac or kotlinc.
 *   2. The file is parsed by the CodeQL front-end.
 *   3. The parsed representation is converted to database tables by
 *      the CodeQL extractor.
 *
 * This class provides CPU and elapsed time information for steps 2 and 3,
 * but not for step 1.
 */
class Compilation extends @compilation {
  /** Gets a textual representation of this element. */
  string toString() {
    exists(string name |
      compilations(this, _, _, name) and
      result = "<compilation " + name + ">"
    )
  }

  /** Gets a file compiled during this invocation. */
  File getAFileCompiled() { result = getFileCompiled(_) }

  /** Gets the `i`th file compiled during this invocation. */
  File getFileCompiled(int i) { compilation_compiling_files(this, i, result) }

  /** Holds if the `i`th file during this invocation was successfully extracted. */
  predicate fileCompiledSuccessful(int i) { compilation_compiling_files_completed(this, i, 0) }

  /** Holds if the `i`th file during this invocation had recoverable extraction errors. */
  predicate fileCompiledRecoverableErrors(int i) {
    compilation_compiling_files_completed(this, i, 1)
  }

  /** Holds if the `i`th file during this invocation had non-recoverable extraction errors. */
  predicate fileCompiledNonRecoverableErrors(int i) {
    compilation_compiling_files_completed(this, i, 2)
  }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * front-end.
   */
  float getFrontendCpuSeconds(int i) { compilation_time(this, i, 1, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * front-end.
   */
  float getFrontendElapsedSeconds(int i) { compilation_time(this, i, 2, result) }

  /**
   * Gets the amount of CPU time spent processing file number `i` in the
   * extractor.
   */
  float getExtractorCpuSeconds(int i) { compilation_time(this, i, 3, result) }

  /**
   * Gets the amount of elapsed time while processing file number `i` in the
   * extractor.
   */
  float getExtractorElapsedSeconds(int i) { compilation_time(this, i, 4, result) }

  /**
   * Gets an argument passed to the extractor on this invocation.
   */
  string getAnArgument() { result = getArgument(_) }

  /**
   * Gets the `i`th argument passed to the extractor on this invocation.
   */
  string getArgument(int i) { compilation_args(this, i, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * compiler.
   */
  float getCompilerCpuSeconds() { compilation_compiler_times(this, result, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the compiler.
   */
  float getCompilerElapsedSeconds() { compilation_compiler_times(this, _, result) }

  /**
   * Gets the total amount of CPU time spent processing all the files in the
   * front-end and extractor.
   */
  float getTotalCpuSeconds() { compilation_finished(this, result, _, _) }

  /**
   * Gets the total amount of elapsed time while processing all the files in
   * the front-end and extractor.
   */
  float getTotalElapsedSeconds() { compilation_finished(this, _, result, _) }

  /**
   * Holds if this is a compilation of Java code.
   */
  predicate isJava() { this instanceof @javacompilation }

  /**
   * Holds if this is a compilation of Kotlin code.
   */
  predicate isKotlin() { this instanceof @kotlincompilation }

  /**
   * Holds if extraction for the compilation started.
   */
  predicate extractionStarted() { compilation_started(this) }

  /**
   * Holds if the extractor terminated normally. Terminating with an exit
   * code indicating that an error occurred is considered normal
   * termination, but crashing due to something like a segfault is not.
   */
  predicate normalTermination() { compilation_finished(this, _, _, _) }

  /**
   * Holds if the extractor succeeded without error.
   */
  predicate extractionSuccessful() { compilation_finished(this, _, _, 0) }

  /**
   * Holds if the extractor encountered recoverable errors.
   */
  predicate recoverableErrors() { compilation_finished(this, _, _, 1) }

  /**
   * Holds if the extractor encountered non-recoverable errors.
   */
  predicate nonRecoverableErrors() { compilation_finished(this, _, _, 2) }
}
