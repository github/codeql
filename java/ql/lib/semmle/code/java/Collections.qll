/**
 * Provides classes and predicates for reasoning about instances of
 * `java.util.Collection` and their methods.
 */

import java

/**
 * Holds if the type `t` is a parameterization of `g`, where the `i`-th type parameter of
 * `g` is instantiated to `arg`.
 *
 * For example, `List<Integer>` parameterizes `List<T>`, instantiating its `0`-th
 * type parameter to `Integer`, while the raw type `List` also parameterizes
 * `List<T>`, instantiating the type parameter to `Object`.
 */
predicate instantiates(RefType t, GenericType g, int i, RefType arg) {
  t = g.getAParameterizedType() and
  arg = t.(ParameterizedType).getTypeArgument(i)
  or
  t = g.getRawType() and
  exists(g.getTypeParameter(i)) and
  arg instanceof TypeObject
}

/**
 * Generalisation of `instantiates` that takes subtyping into account:
 *
 * - `HashSet<Integer>` indirectly instantiates `Collection` (but also `HashSet` and `Set`),
 *   with the `0`-th type parameter being `Integer`;
 * - a class `MyList extends ArrayList<Runnable>` also instantiates `Collection`
 *   (as well as `AbstractList`, `AbstractCollection` and `List`), with the `0`-th type
 *   parameter being `Runnable`;
 * - the same is true of `class MyOtherList<T> extends ArrayList<Runnable>` (note that
 *   it does _not_ instantiate the type parameter to `T`);
 * - a class `MyIntMap<V> extends HashMap<Integer, V>` instantiates `Map` (among others)
 *   with the `0`-th type parameter being `Integer` and the `1`-th type parameter being `V`.
 */
pragma[nomagic]
predicate indirectlyInstantiates(RefType t, GenericType g, int i, RefType arg) {
  instantiates(t, g, i, arg)
  or
  exists(RefType sup |
    t.extendsOrImplements(sup) and
    indirectlyInstantiates(sup, g, i, arg)
  )
}

/** A reference type that extends a parameterization of `java.util.Collection`. */
class CollectionType extends RefType {
  CollectionType() {
    this.getSourceDeclaration().getASourceSupertype*().hasQualifiedName("java.util", "Collection")
  }

  /** Gets the type of elements stored in this collection. */
  RefType getElementType() {
    exists(GenericInterface coll | coll.hasQualifiedName("java.util", "Collection") |
      indirectlyInstantiates(this, coll, 0, result)
    )
  }
}

/** A method declared in a collection type. */
class CollectionMethod extends Method {
  CollectionMethod() { this.getDeclaringType() instanceof CollectionType }

  /** Gets the type of elements of the collection to which this method belongs. */
  RefType getReceiverElementType() {
    result = this.getDeclaringType().(CollectionType).getElementType()
  }
}

/** The `size` method on `java.util.Collection`. */
class CollectionSizeMethod extends CollectionMethod {
  CollectionSizeMethod() { this.hasName("size") and this.hasNoParameters() }
}

/** A method that mutates the collection it belongs to. */
class CollectionMutator extends CollectionMethod {
  CollectionMutator() {
    pragma[only_bind_into](this).getName().regexpMatch("add.*|remove.*|push|pop|clear")
  }
}

/** A method call that mutates a collection. */
class CollectionMutation extends MethodAccess {
  CollectionMutation() { this.getMethod() instanceof CollectionMutator }

  /** Holds if the result of this call is not immediately discarded. */
  predicate resultIsChecked() { not this instanceof ValueDiscardingExpr }
}

/** A method that queries the contents of a collection without mutating it. */
class CollectionQueryMethod extends CollectionMethod {
  CollectionQueryMethod() {
    pragma[only_bind_into](this).getName() = ["contains", "containsAll", "get", "size", "peek"]
  }
}

/** A `new` expression that allocates a fresh, empty collection. */
class FreshCollection extends ClassInstanceExpr {
  FreshCollection() {
    this.getConstructedType() instanceof CollectionType and
    this.getNumArgument() = 0 and
    not exists(this.getAnonymousClass())
  }
}
