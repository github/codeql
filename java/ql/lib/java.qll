/** Provides all default Java QL imports. */

import Customizations
import semmle.code.FileSystem
import semmle.code.Location
import semmle.code.Unit
import semmle.code.java.Annotation
import semmle.code.java.Compilation
import semmle.code.java.CompilationUnit
import semmle.code.java.ControlFlowGraph
import semmle.code.java.Dependency
import semmle.code.java.Element
import semmle.code.java.Exception
import semmle.code.java.Expr
import semmle.code.java.GeneratedFiles
import semmle.code.java.Generics
import semmle.code.java.Import
import semmle.code.java.J2EE
import semmle.code.java.Javadoc
import semmle.code.java.JDK
import semmle.code.java.JDKAnnotations
import semmle.code.java.JMX
import semmle.code.java.KotlinType
import semmle.code.java.Member
import semmle.code.java.Modifier
import semmle.code.java.Modules
import semmle.code.java.Package
import semmle.code.java.Statement
import semmle.code.java.Type
import semmle.code.java.UnitTests
import semmle.code.java.Variable
import semmle.code.java.controlflow.BasicBlocks
import semmle.code.java.metrics.MetricCallable
import semmle.code.java.metrics.MetricElement
import semmle.code.java.metrics.MetricField
import semmle.code.java.metrics.MetricPackage
import semmle.code.java.metrics.MetricRefType
import semmle.code.java.metrics.MetricStmt
import semmle.code.xml.Ant
import semmle.code.xml.XML
