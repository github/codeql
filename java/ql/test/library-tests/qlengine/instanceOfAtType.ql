import java

from Class c, Member m
where
  c.hasQualifiedName("", "Tst") and
  c.getAMember() = m and
  m instanceof @field
select m
