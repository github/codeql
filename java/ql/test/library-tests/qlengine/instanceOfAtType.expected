| Tst.java:2:8:2:12 | field |
